[Permission] permission domain app&data 계층 구현

vwx12345 opened the pull request:
### 💻 작업 내용
- permission 도메인의 application 과 data access 계층을 구현했습니다.

### ✨ 리뷰 포인트
- 

### 📝 메모
- jpa 관련도 commit 했는데 이로 인해 발생한 conflict 는 제가 해결하겠습니다.

### 🎯 관련 이슈
closed #49

vwx12345 committed:
feat(permission): permission command application 영역 구현
- PermissionCommandRepository 구현
- PermissionCommandService 구현
- DelegateRoleCommand(dto) 구현

vwx12345 committed:
feat(permission): UserFetchService 구현
- Permission 도메인에서 User 정보를 가져오기 위한 서비스 구현

vwx12345 committed:
feat(permission): permission query application 영역 구현
- PermissionQueryRepository 구현
- PermissionQueryService 구현

vwx12345 committed:
feat(permission): UserFetchService 임시 구현
- 우선 UserId를 1L을 반환하도록 임시로 구현

vwx12345 committed:
feat(permission): 데이터 접근용 entity 구현 및 build.gradle 수정
- jpa를 사용하기 위해 build.gradle 수정
- jpa를 사용해 entity 구현

vwx12345 committed:
feat(permission): permission Command repository 구현
- PermissionCommandJpaRepository 구현
- PrmissionCommandRepositoryImpl 구현

vwx12345 committed:
feat(permission): permission Query repository 구현
- PermissionQueryJpaRepository 구현
- PermissionQueryRepositoryImpl 구현

vwx12345 committed:
test(permission): test 코드 작성

vwx12345 committed:
docs(permission): 주석 추가 및 spotless 적용

codingbaraGo commented:
ci 테스트 통과 후 리뷰어 다시 등록해주세요
일단 빼놓겠습니다

vwx12345 committed:
fix(permission): 의존성 있는 테스트 제거

codingbaraGo commented on src/main/java/com/simpleboard/board/permission/infrastructure/UserFetchServiceImpl.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class UserFetchServiceImpl implements UserFetchService {
+
+  @Override
+  public Long getUserIdByNickname(String nickname) {
+    return 1L;
+  }
```
테스트를 위해 구현체를 구현한것이라면, 테스트에서  mock을 사용하는게 좋지 않을까요?
혹은 해당 구현체를 테스트 패키지에 넣고, 클래스 이름에 테스트용 클래스라는 것을 명시하여도 좋구요.

vwx12345 replied:
> 테스트를 위해 구현체를 구현한것이라면, 테스트에서 mock을 사용하는게 좋지 않을까요? 혹은 해당 구현체를 테스트 패키지에 넣고, 클래스 이름에 테스트용 클래스라는 것을 명시하여도 좋구요.

구현체가 없어 Autowire 에러가 발생하여 임시로 기본 구현체를 생성해두었습니다.
실제 테스트에서는 해당 컴포넌트를 직접 사용하는 것이 아니기 때문에, 테스트 코드에서는 Mock을 주입해 사용하고 있습니다.
혹은 말씀 주신 것처럼 테스트용 구현체로 명확히 분리하는 방향도 고려해보겠습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/permission/infrastructure/entity/PermissionPolicyEntity.java:
```diff
@@ -0,0 +1,45 @@
+
+  public PermissionPolicy toDomain() {
+    List<ManagerAssignment> domainAssignments =
+        managerAssignments.stream().map(ManagerAssignmentEntity::toDomain).toList();
```
stream().toList()로 생성된 list는 불변리스트라 값을 수정할 수 없게 됩니다.
도메인 엔티티를 이렇게 생성하면 ManagerAssignment를 추가, 삭제할 수 없게 되어버려요.

vwx12345 replied:
> stream().toList()로 생성된 list는 불변리스트라 값을 수정할 수 없게 됩니다. 도메인 엔티티를 이렇게 생성하면 ManagerAssignment를 추가, 삭제할 수 없게 되어버려요.

JPA Entity로서 불변성을 지향하기 위해 stream().toList()를 사용해 리스트를 생성했습니다.
다만, PermissionPolicy 내에서 해당 리스트를 조작해야 하는 경우가 있어, 내부에서는 명시적으로 new ArrayList<>(...)로 변환해 사용하고 있어 동작에는 문제가 없습니다.
추후 혼란을 줄이기 위해, 생성 시점부터 가변 리스트로 변경하거나, 생성 메서드 명명이나 주석으로 불변 리스트임을 명확히 표기하는 것도 고려하겠습니다.

codingbaraGo commented on src/test/java/com/simpleboard/board/permission/application/command/PermissionCommandServiceTest.java:
```diff
@@ -0,0 +1,76 @@
+    RoleName roleName = RoleName.BOARD_ADMIN;
+    DelegateRoleCommand command = new DelegateRoleCommand(creatorId, toNickname, roleName);
+    when(permissionCommandRepository.getByBoardId(boardId)).thenReturn(permissionPolicy);
+    when(userFetchService.getUserIdByNickname(toNickname)).thenReturn(toUserId);
```
테스트에서 JPA repository가 아닌, command repository 인터페이스를 mock으로 대체하는데, 이로 인해 도메인 엔티티 <-> JPA 엔티티 간의 convert 과정이 생략되네요.
command repository 구현체 내부에서 사용하는 JPA repository에 mock을 사용해서 JPA엔티티를 반환하도록 하면 convert 과정까지 검증할 수 있지 않을까 합니다.

codingbaraGo replied:
만약 서비스 클래스 테스트에서는 서비스 로직만 테스트하고 싶으시다면, 
converter 테스트에서 엔티티 변환 후 여러 도메인 메서드를 사용하며 변환된 도메인 엔티티 클래스가 정상적으로 작동하는지 체크해주는것도 방법인것 같습니다.

vwx12345 replied:
> 테스트에서 JPA repository가 아닌, command repository 인터페이스를 mock으로 대체하는데, 이로 인해 도메인 엔티티 <-> JPA 엔티티 간의 convert 과정이 생략되네요. command repository 구현체 내부에서 사용하는 JPA repository에 mock을 사용해서 JPA엔티티를 반환하도록 하면 convert 과정까지 검증할 수 있지 않을까 합니다.

테스트에서 CommandRepository를 mock으로 대체하다 보니, 자연스럽게 JPA Entity ↔ 도메인 Entity 간의 convert 과정이 생략되는 구조가 되었습니다.
이 과정에서 변환 로직의 정상 동작 여부가 테스트되지 않는 점이 다소 아쉽다고 느꼈고,
CommandRepository 구현체 내부에서 사용하는 JPA Repository에 mock을 주입하여, JPA Entity를 반환하게 하면 실제 convert 로직까지 포함된 테스트를 구성할 수 있겠다는 아이디어를 떠올렸습니다.

vwx12345 replied:
> 만약 서비스 클래스 테스트에서는 서비스 로직만 테스트하고 싶으시다면, converter 테스트에서 엔티티 변환 후 여러 도메인 메서드를 사용하며 변환된 도메인 엔티티 클래스가 정상적으로 작동하는지 체크해주는것도 방법인것 같습니다.

사실 저도 서비스 테스트에선 서비스 로직만 검증하는 게 맞다고 생각합니다.
다만 현재는 전체 흐름을 빠르게 확인하기 위해, 변환 로직까지 포함한 형태로 테스트를 작성했습니다.
이후에는 변환 책임을 명확히 분리해서, converter에 대한 단위 테스트를 따로 구성하고, 서비스 테스트에서는 로직 자체만 검증하는 방향으로 리팩토링할 계획입니다.

codingbaraGo reviewed (changes requested):
git 기능 중 rebase, soft reset, amend 라는 기능들이 있더라구요.
해당 기능들을 사용하면 커밋을 보다 보기좋고 깔끔하게 구성하여 push를 할 수 있습니다.
저도 지금 커밋을 깔끔하게 만드는 연습을 하는 중입니다.

현업에서는 코드를 볼 때 커밋단위로 보는 경우가 많기 때문에 커밋을 보는사람이 보기 좋게 구성하는것도 중요한 요소중 하나라고 합니다.
커밋을 작성 후 해당 커밋의 코드에 대한 수정사항이 발생한다면, git의 기능들을 사용하여 outdated 코드가 들어있는 커밋이 없도록 해보는건 어떨까요?

이 PR 말고 다음 PR 부터는 여러 기능들을 한번씩 써보시면 좋을것같아요.

vwx12345 committed:
refactor(permission): PermissionCommandRepository 이름 변경
- PermissionCommandRepository를 도메인 영역으로 옮기면서 PermissionRepository로 이름을 바꿨습니다.
- PermissionCommandRepository를 사용하던 클래스에서 PermissionRepository로 변경했습니다.

vwx12345 committed:
docs(permission): 주석 변경

vwx12345 committed:
test(permission): permissionCommandJpaRepository mocking하여 Test

vwx12345 committed:
chore(permission): spotless apply

vwx12345 commented:
> git 기능 중 rebase, soft reset, amend 라는 기능들이 있더라구요. 해당 기능들을 사용하면 커밋을 보다 보기좋고 깔끔하게 구성하여 push를 할 수 있습니다. 저도 지금 커밋을 깔끔하게 만드는 연습을 하는 중입니다.
> 
> 현업에서는 코드를 볼 때 커밋단위로 보는 경우가 많기 때문에 커밋을 보는사람이 보기 좋게 구성하는것도 중요한 요소중 하나라고 합니다. 커밋을 작성 후 해당 커밋의 코드에 대한 수정사항이 발생한다면, git의 기능들을 사용하여 outdated 코드가 들어있는 커밋이 없도록 해보는건 어떨까요?
> 
> 이 PR 말고 다음 PR 부터는 여러 기능들을 한번씩 써보시면 좋을것같아요.

항상 rebase만 했었는데 다른 부분도 생각해봐야겠네요. 조언 감사합니다!

codingbaraGo reviewed (approved):
고생하셨습니다.
전체적으로 다 좋은데 커밋 메시지 작성하실 때
인터페이스 클래스를 생성하고 정의하는 경우에는 "구현"이라는 단어보다는 "정의"가 더 직관적이지 않을까 제안드립니다.
